openpnp/openpnp
# Feature/Automatic Machine Calibration using Issues & Solutions - Round 3

# Description
This roundup of enhancements include the following:

- Merged develop into test
- Used @tonyluken 's new SimpleGraph improvements to better visualize backlash error data.
    
    ![New Backlash Calibration Graphs](https://user-images.githubusercontent.com/9963310/130335118-e0eaf9dc-5786-4803-8ed4-f59e73bf21fc.png)

- Introduces `UiUtils.confirmMoveToLocationAndAct()` for UI actions that need to perform moves by the machine, that are possibly unexpected by the user. So this wrapper asks for confirmation and handles a disabled machine, threading issues and exception message boxes. Refactored `ReferencePushPullFeederConfigurationWizard` pre pipeline editing move (where this was extracted). Added a `ReferenceNozzleTipCalibrationWizard` pre pipeline editing move.
- Fix null pointer exception due to 3D UPP configuration error in `CameraView`.
- Move the `Head.isInsideSoftLimits()` to `Movable.isReachable()`, as it is actually not Head business.
- When setting head offsets on a `ReferenceNozzle`, update other, dependent head offsets, including bottom camera locations (which are head offsets internally). This allows recalibration of nozzle head offsets without invalidating other calibrations.
- Improve the Auto-Focus to always use the same servo direction. Add a small retract distance before each iteration. This gets rid of any uncompensated backlash in the axis (which is typically the case with Z axes that are used to auto-focus).
- Increase the auto-calibration Auto-Focus focal depth.
- Fix a bug/ambiguity in unit vector LengthUnits handling.
- When Issues & Solutions nozzle solutions are applied that have fewer nozzles than before, delete surplus actuators (as it already did with nozzles and axes).
- Allow for actuators on Head or Machine to be permutated with Up/Down arrow buttons in the Machine Setup tree, same as Drivers, Axes.
- Make sure solutions are presented, when absolutely needed, even when marked as "solved".
- Added this to camera preliminary calibrations in case Units per Pixels are missing.
- Issues & Solutions Visual Homing calibration will now also create the "FIDUCIAL-HOME" package and part, and set its diameter based on interactive diameter setting. A `DetectCircularSymmetry` stage based vision pipeline is automatically assigned, providing tuning free fiducial vision.
- Add `LengthProperty` to `Solution.Issue` and `IssuePanel`
- Add `acceptableTolerance` to `ReferenceControllerAxis` and `ReferenceControllerAxisConfigurationWizard`.
- Use` acceptableTolerance` in `CalibrationSolutions` backlash compensation.
- Improve `calibrateAxisBacklash` to consider both directed and reversed moves. Add random move test.
- `SimpleGraph` should not interpolate between points that have no line shown (scatter plot). Instead, snap to the nearest data point.

- Bug-fix (unit tests only): Prevent `Configuration.getMachine()` not yet ready race condition in `SimulationModeMachine`.

# Justification
Ongoing testing in the testing version.

# Instructions for Use
Updated the Wiki:
https://github.com/openpnp/openpnp/wiki/Issues-and-Solutions#kinematics-milestone

# Implementation Details
1. Tested both in simulation and on the machine.
2. Did follow the [coding style](https://github.com/openpnp/openpnp/wiki/Developers-Guide#coding-style).
3. Small changes in the `org.openpnp.spi` or `org.openpnp.model` packages. `Head.isInsideSoftLimits()` to `Movable.isReachable()`.
4. Successful `mvn test` before submitting the Pull Request.
